Logic/item_detection.py

```diff
@@ -0,0 +1,196 @@
+'''Item Detection Module'''
+
+import os
+import numpy as np
+import cv2
+
+from skimage.measure import compare_ssim as ssim
+
+from Config.eight_ball_lookup import EightBallLookup
+
+class ItemDetection:
+    '''Responsible for detecting game balls and holes'''
+
+    lookup = EightBallLookup()
+
+    def identify_parameters(self, for_holes, for_balls):
+        '''Responsible iterating through a number of images and saving the image results based on different parameter values'''
+
+        for ball_folder in os.listdir(self.lookup.training_path):
+            ball_path = os.path.join(self.lookup.training_path, ball_folder)
+
+            if (os.path.isfile(ball_path) and ball_path.lower().endswith(('.png', '.jpg', '.jpeg'))):
+                if for_holes:
+                    if not os.path.exists(self.lookup.hole_training_path):
+                        os.makedirs(self.lookup.hole_training_path)
+
+                    self.find_hole_parameters(ball_path)
+
+                if for_balls:
+                    if not os.path.exists(self.lookup.ball_training_path):
+                        os.makedirs(self.lookup.ball_training_path)
+
+                    hole_positions = self.detect_holes(ball_path)
+                    self.find_ball_parameters(ball_path, hole_positions)
+
+    def find_hole_parameters(self, image_path):
+        '''Iterate through a range of values and saving each result in the aim of identifying the best parameters for hole detection'''
+
+        for param2 in range(13, 17, 1):
+            for param1 in range(10, 310, 10):
+                print('Hole Parameters:', param1, param2)
+
+                rgb_image = cv2.imread(image_path)
+                gray_image = cv2.cvtColor(rgb_image, cv2.COLOR_BGR2GRAY)
+
+                holes = cv2.HoughCircles(gray_image, cv2.HOUGH_GRADIENT, 1, 10, param1=param1, param2=param2, minRadius=20, maxRadius=21)
+
+                if holes is not None:
+                    holes = np.round(holes[0, :]).astype("int")
+
+                    for (x_position, y_position, radius) in holes:
+                        cv2.circle(rgb_image, (x_position, y_position), 2, (15, 10, 25), 3)
+                        cv2.circle(rgb_image, (x_position, y_position), radius, (25, 10, 255), 3)
+
+                cv2.imwrite(self.lookup.hole_training_path + str(param2) + '_' + str(param1) + '.jpg', rgb_image)
+
+    def find_ball_parameters(self, image_path, hole_positions):
+        '''Iterate through a range of values and saving each result in the goal of identify the best parameters for ball detection'''
+
+        min_x = min(hole_positions, key=lambda t: t[0])[0] + 10
+        min_y = min(hole_positions, key=lambda t: t[1])[1] + 10
+        max_x = max(hole_positions, key=lambda t: t[0])[0] - 10
+        max_y = max(hole_positions, key=lambda t: t[1])[1] - 10
```

Comment:
Add comment to signify that these values correspond to the corners

Reply:
Done